### PR TITLE
Use scheduling package in filterOutSchedulable processor

### PR DESCRIPTION
**Pull request by BigDarkClown**
#### Which component this PR applies to?

cluster-autoscaler

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR separates the binpacking logic from the filterOutSchedulablePodsListProcessor. This enables reausability of this code and allows other places in CA to use it.

#### Does this PR introduce a user-facing change?

NONE

### CODE REVIEW

File: cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go [hunk 1]
```diff
@@ -93,110 +92,135 @@ func (p *filterOutSchedulablePodListProcessor) Process(
...
-	var schedulablePods []*apiv1.Pod
-	for _, x := range allUnschedulablePods {
-		if _, found := podsStillUnschedulableMap[x]; !found {
-			schedulablePods = append(schedulablePods, x)
+// ClearCache clears the cached hints and other auxiliary data to avoid memory overflow
+func (f *SchedulablePodFilter) ClearCache() {
+	// clear hints for schedulable pods
+	for hintedPodUID := range f.schedulablePodsNodeHints {
+		if !f.processedPods[hintedPodUID] {
+			delete(f.schedulablePodsNodeHints, hintedPodUID)
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go]: That's a good point, I've scrapped the previous implementation and moved to the scheduling package one.

File: cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go [hunk 2]
```diff
@@ -97,94 +94,26 @@ func (p *filterOutSchedulablePodListProcessor) CleanUp() {
...
-		if err == nil {
-			klog.V(4).Infof("Pod %s.%s marked as unschedulable can be scheduled on node %s. Ignoring"+
-				" in scale up.", pod.Namespace, pod.Name, nodeName)
-			if err := clusterSnapshot.AddPod(pod, nodeName); err != nil {
-				return nil, err
-			}
-			// Store hint for pod placement
-			p.schedulablePodsNodeHints[pod.UID] = nodeName
-		} else {
+		_, err := p.schedulingSimulator.TrySchedulePods(clusterSnapshot, []*apiv1.Pod{pod}, scheduling.ScheduleAnywhere)
```
Comment (x13n) [inline cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go:106]: Maybe instead of looping over pods here it would be possible to return some `SchedulingStatus` struct as a first argument and have unschedulable pods in it as a field?

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go [hunk 3]
```diff
@@ -65,26 +66,37 @@ func (s *HintingSimulator) TrySchedulePods(clusterSnapshot clustersnapshot.Clust
...
-		return isNodeAcceptable(nodeInfo.Node().Name)
-	})
+func (s *HintingSimulator) schedulePod(clusterSnapshot clustersnapshot.ClusterSnapshot, pod *apiv1.Pod, isNodeAcceptable func(*schedulerframework.NodeInfo) bool) (string, error) {
+	if _, found := s.unschedulablePods.Get(pod); found {
+		return "", fmt.Errorf("failed to find place for %s/%s based on caching", pod.Namespace, pod.Name)
+	}
+
+	newNodeName, err := s.predicateChecker.FitsAnyNodeMatching(clusterSnapshot, pod, isNodeAcceptable)
 	if err != nil {
+		s.unschedulablePods.Set(pod, nil)
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go]: Agreed, I changed the interface and name of the struct to better reflect what it does.

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go [hunk 4]
```diff
@@ -98,9 +110,15 @@ func (s *HintingSimulator) schedulePod(clusterSnapshot clustersnapshot.ClusterSn
 // DropOldHints drops old scheduling hints.
 func (s *HintingSimulator) DropOldHints() {
 	s.hints.DropOld()
+	s.unschedulablePods = NewPodSchedulableMap()
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go]: You're right, I moved this to the start of `TrySchedulePods `

File: cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go (hunk 2 above)
Comment (BigDarkClown) [reply cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go]: I've added a `scheduling.Status` struct that contains:
- the scheduled pod
- the node name

I've also needed to change the `TrySchedulePods` logic - up until now it failed if there was an unexpected error OR if some pod was unschedulable. Now it fails only on unexpected errors and returns the subset of scheduled pods if it does not manage to schedule all.

This required some changes around calls to `TrySchedulePods`, but reproducing the previous behaviour can be done just by comparing the length of the pod list vs the length of the scheduled pods list.

Comment (x13n) [reply cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go]: Comparing list lengths makes sense from the correctness point of view, but I'd consider having a bool param in scheduling simulator (ctor) whether to error out early on first scheduling error - this may have significant impact on scale down performance.

Comment (x13n) [reply cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go]: (Alternatively: a param to `TrySchedulePods`)

Comment (BigDarkClown) [reply cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go]: Added `scheduleAll` param to the `TrySchedulePods` - this should help from the performance perspective. 

I don't think however it would be good to return an error if we want to schedule all. Having a function which return two different formats depending on the boolean flag is confusing. So I left the length comparison I previously added, and just added a `break` to the `TrySchedulePods`.

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go [hunk 5]
```diff
@@ -44,53 +50,73 @@ func NewHintingSimulator(predicateChecker predicatechecker.PredicateChecker) *Hi
 
 // TrySchedulePods attempts to schedule provided pods on any acceptable nodes.
 // Each node is considered acceptable iff isNodeAcceptable() returns true.
-// Returns a list of nodes that were chosen or an error if scheduling any of the pods failed.
+// Returns a list of scheduled nodes or an error if an unexpected error occurs.
```
Comment (towca) [inline cluster-autoscaler/simulator/scheduling/hinting_simulator.go:53]: scheduled pods*

File: cluster-autoscaler/simulator/scheduling/similar_pods.go [hunk 6]
```diff
@@ -36,56 +36,55 @@ import (
 // calculated.
 // To decide if two pods are similar enough we check if they have identical label
 // and spec and are owned by the same controller. The problem is the whole
-// PodSchedulableInfo struct is not hashable and keeping a list and running deep
+// SimilarPodsSchedulingInfo struct is not hashable and keeping a list and running deep
 // equality checks would likely also be expensive. So instead we use controller
 // UID as a key in initial lookup and only run full comparison on a set of
 // podSchedulableInfos created for pods owned by this controller.
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/similar_pods.go:42]: Done

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go [hunk 7]
```diff
@@ -44,53 +50,73 @@ func NewHintingSimulator(predicateChecker predicatechecker.PredicateChecker) *Hi
...
+	if hintedNode, hasHint := s.hints.Get(hk); hasHint {
 		if err := s.predicateChecker.CheckPredicates(clusterSnapshot, pod, hintedNode); err == nil {
-			klog.V(4).Infof("Pod %s/%s can be moved to %s", pod.Namespace, pod.Name, hintedNode)
-			if err := clusterSnapshot.AddPod(pod, hintedNode); err != nil {
-				return "", fmt.Errorf("Simulating scheduling of %s/%s to %s return error; %v", pod.Namespace, pod.Name, hintedNode, err)
-			}
 			s.hints.Set(hk, hintedNode)
-			return hintedNode, nil
+
+			nodeInfo, err := clusterSnapshot.NodeInfos().Get(hintedNode)
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go:94]: Assume the scenario when pod A is able to get scheduled only on node B. We detect it once and cache it. Then somebody runs the algorithm again, this time with node B being non-acceptable. This can happen e.g. in scale-down, when non-acceptable nodes are different each run.
If we just ignored the hint because node B is non-acceptable, then we would remove this data from cache. This is not ideal - consecutive call can have B as acceptable, but the hint will already be deleted.

To recap - thanks to this single predicate run we can potentially avoid O(N) predicate checks in the future. This is a small optimalization, but can potentially help.

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go [hunk 8]
```diff
@@ -44,53 +50,73 @@ func NewHintingSimulator(predicateChecker predicatechecker.PredicateChecker) *Hi
...
 
-func (s *HintingSimulator) schedulePod(clusterSnapshot clustersnapshot.ClusterSnapshot, pod *apiv1.Pod, isNodeAcceptable func(string) bool) (string, error) {
-	newNodeName, err := s.predicateChecker.FitsAnyNodeMatching(clusterSnapshot, pod, func(nodeInfo *schedulerframework.NodeInfo) bool {
-		return isNodeAcceptable(nodeInfo.Node().Name)
-	})
-	if err != nil {
-		return "", fmt.Errorf("failed to find place for %s/%s", pod.Namespace, pod.Name)
+func (s *HintingSimulator) findNode(clusterSnapshot clustersnapshot.ClusterSnapshot, pod *apiv1.Pod, isNodeAcceptable func(*schedulerframework.NodeInfo) bool) (string, error) {
+	if s.similarPods.IsSimilarUnschedulable(pod) {
+		klog.V(4).Infof("failed to find place for %s/%s based on similar pods scheduling", pod.Namespace, pod.Name)
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go:109]: I don't really agree - these are logs about pods not being able to move somewhere, I don't think they necessarily need to be in the same place. 

Also logging this in `TrySchedulePods` would require returning some reasoning from `findNode`. Using error would IMHO be the only reasonable way, but that would change the behaviour and error handling - I'm not a fan of this.

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go [hunk 9]
```diff
@@ -100,7 +126,12 @@ func (s *HintingSimulator) DropOldHints() {
 	s.hints.DropOld()
 }
 
+// OverflowingControllerCount returns number of overflowing controllers
+func (s *HintingSimulator) OverflowingControllerCount() int {
+	return s.similarPods.OverflowingControllerCount()
```
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go:131]: Makes sense, added this as a returned value for `TrySchedulePods`.

[the remark this answers is not in the document]
Comment (towca) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go]: Could you also make it not a field of HintingSimulator?

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (towca) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go:111]: Sure, not going to argue here, doesn't matter much :P

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (towca) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go:96]: Ack, makes sense

File: cluster-autoscaler/simulator/scheduling/hinting_simulator.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (BigDarkClown) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go]: Done, also I changed the `NewSimilarPodsScheduling` associated functions to work on pointers.

[the remark this answers is not in the document]
Comment (towca) [reply cluster-autoscaler/simulator/scheduling/hinting_simulator.go]: Thx, looks much better!
